File: Solutions/Python/binary_tree_level_order_traversal.py
"""
    [MEDIUM]
    102. Binary Tree Level Order Traversal

    Concepts:
    - binary tree
    - breadth-first search

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 18.47 MB  [Beats 60.04%]
"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []
            
        levels = []

        queue = deque([root])
        while queue:
            n = len(queue)
            level = []

            for _ in range(n):
                curr = queue.popleft()
                level.append(curr.val)

                if curr.left:
                    queue.append(curr.left)
                if curr.right:
                    queue.append(curr.right)

            levels.append(level)
        
        return levels

File: Solutions/Cpp/binary_tree_level_order_traversal.cpp
/*
    [MEDIUM]
    102. Binary Tree Level Order Traversal

    Concepts:
    - binary tree
    - breadth-first search

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 16.90 MB  [Beats 93.68%]
*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        std::vector<std::vector<int>> levels = {};
        std::vector<int> level = {};

        if (!root)
            return levels;

        std::queue<TreeNode*> q;
        q.push(root);
        int n;
        TreeNode* curr;

        while (!q.empty()) {
            n = q.size();
            level.clear();

            for (int _ = 0; _ < n; _++) {
                curr = q.front();
                q.pop();
                level.push_back(curr->val);

                if (curr->left)
                    q.push(curr->left);
                if (curr->right)
                    q.push(curr->right);
            }

            levels.push_back(level);
        }

        return levels;
    }
};